woocommerce/google-listings-and-ads
# Shipping rates UI changes: Change `saveShippingRates` into custom hook

### Changes proposed in this Pull Request:

Closes #1287.

In this PR, we changed the `saveShippingRates` wp.data actions into custom hook. This addresses the points noted in issue #1287:

1. We add `upsertShippingRates` and `deleteShippingRates` wp.data actions, and use `saveShippingRates` custom hook that wraps around them.
2. `upsertShippingRates` will only be called for changed or newly added shipping rates.
3. We have a `getDeleteIds` function to encapsulate the logic of how to get the IDs to be deleted.
4. No try/catch block in `upsertShippingRates` and `deleteShippingRates` wp.data actions and in `saveShippingRates` custom hook, so that the error can propagate up and be caught by the components to display the necessary error message based on the context.

### Screenshots:

📹 Demo video with my voice:

https://user-images.githubusercontent.com/417342/157846888-e14921e8-2ec3-4ccb-84da-3af1059e9d06.mov


### Detailed test instructions:

You can test this in Setup MC or Edit Free Listings.

1. Open browser dev tools to observe network requests. It is recommended for you to have the Redux Dev Tools to observe wp.data store changes.
2. Go to shipping rate section. Add, edit, or delete shipping rates.
    - Note: to delete shipping rate, you may have to modify the target audience countries.
3. Click on "Continue" / "Save changes" button. In the network panel, there should be a delete shipping rate request for deleted shipping rates, and upsert shipping rate request for added or edited shipping rates. When the requests are done, it should be reflected in the wp.data store.
4. Without reloading the page, repeat step 2 and 3 above for a few more times. This is to test the comparison between the shipping rates from form data and wp.data store.
5. Reload the page. The shipping rates should be the same as before the reload.


### Changelog entry
